### PR TITLE
doc/man/8: add documentation about read balancer

### PR DESCRIPTION
**Pull request by ljflores**
Documentation for the read balancer feature introduced in Reef.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by anthonyeleven** [comment]
LGTM from a docs perspective, but I have a functional question:  if read balancing is done by managing primary affinity, is drive capacity an input?  And will the module do the right thing when cluster topology changes?

**Comment by ljflores** [comment]
> LGTM from a docs perspective, but I have a functional question: if read balancing is done by managing primary affinity, is drive capacity an input?

Hey Anthony, the current primary balancing algorithm assumes the cluster has OSDs of the same size. So, at the moment, drive capacity (I think this is what you mean by "drive capacity"?) is not an input, although this considering drives of different sizes will be an improvement in a future version of the primary balancer. The current algorithm is ideal for balancing small, homogeneous clusters, i.e. ODF clusters.

A user may still run the current balancer on a cluster with different drive sizes, but the optimization may not be the the *best* optimization.

Primary affinity is indeed factored into the current algorithm, so if a user sets an OSD to 0 primary affinity, this will be honored by the read balancer, and no primaries will be assigned to that OSD.

I will add another commit to this PR explaining these nuances.

> And will the module do the right thing when cluster topology changes?

If the topology changes, i.e. as a result of the pg autoscaler, the user can simply rerun the read balancer on their changed OSD map. They may want to check how the `read_balance_score` has changed on each pool by running `ceph osd pool ls detail`.

It is another future improvement to add the read balancer to the current balancer manager module so it is always running in the background, but for now, the user will have to rebalance manually using the offline osdmaptool.

**Comment by ljflores** [comment]
I marked this PR as a draft since @JoshSalomon and I need to add a commit explaining the high level design (which will answer the two questions you asked.

Thanks for the feedback!

**Comment by ljflores** [comment]
@anthonyeleven forgot to tag you. :)

**Comment by anthonyeleven** [comment]
Groovy, thanks for the info.  In a previous life I had a frankencluster that started with a bad topology and got worse with each expansion.  Suboptimal CRUSH rule, mix of three different drive sizes.  Experiments with PA were interesting, but the two questions above were concerns at the time.  Calculating an optimal set of PA weights did not seem trivial, and it seemed that the loss or addition of an OSD might perturb those calculations such that a given OSD might end up being worse off than it was originally.

**Comment by ljflores** [comment]
Rendered here:

https://ceph--51499.org.readthedocs.build/en/51499/man/8/osdmaptool/

**Review by anthonyeleven — APPROVED** [review]

File: doc/dev/balancer-design.rst [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+of the read balancer.
+
+Future Possible Improvements
+----------------------------
+
+Ideas for future versions include:
+
+1. Improve behavior for heterogeneous OSDs in a pool
+2. Offer read balancing as an online option to the balancer manager module
+3. Handle workload balancing based on real load statistics of the OSDs
```
Comment (ljflores) [inline doc/dev/balancer-design.rst:66]: Remove this

File: doc/dev/balancer-design.rst [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+How to Balance Reads in Ceph
+----------------------------
+See [operating the read balancer] for more information.
+
+Also, see the Cephalocon 2023 talk `New Read Balancer in Ceph <https://www.youtube.com/watch?v=AT_cKYaQzcU/>`_ for a demonstration of the offline version
+of the read balancer.
+
+Future Possible Improvements
+----------------------------
+
```
Comment (ljflores) [inline doc/dev/balancer-design.rst:61]: Remove

**Comment by ljflores** [comment]
Rendered here:

https://ceph--51499.org.readthedocs.build/en/51499/dev/balancer-design/
https://ceph--51499.org.readthedocs.build/en/51499/man/8/osdmaptool/
https://ceph--51499.org.readthedocs.build/en/51499/rados/operations/pools/
https://ceph--51499.org.readthedocs.build/en/51499/rados/operations/upmap/
https://ceph--51499.org.readthedocs.build/en/51499/rados/operations/read-balancer/
https://ceph--51499.org.readthedocs.build/en/51499/rados/operations/balancer-module/

**Review by JoshSalomon — APPROVED** [review]

**Comment by zdover23** [comment]
03 Jun 2024:

This information is present in Squid and in Reef, but not in Quincy. As far as I know, this is as it should be.

**Pull request merged**